### PR TITLE
changed colony buildings to use ShortestPath rather than JumpsBetween, 

### PR DESCRIPTION
**Pull request by Dilvish-fo**
using a ship speed of 80 as reference, and with a slight exponential reduction meant to roughly account for very large empires being likely to have faster ships

this content requires recent updates to the FO executables

### CODE REVIEW

**Comment by Dilvish-fo** [comment]
It's hardly fair to complain this is not exact when it was explicitly described as  **_roughly**_ accounting for increasing ship speeds.   I did consider a fairly exact correspondence by taking a max speed of all ships owned by the empire; I passed on that to streamline the calculation but it might not really be all that costly of a calculation so we could give that a try.  I don't like the idea of something that just scales up over time regardless of what the empire has accomplished or not accomplished.  The distance involved at least was very likely to have a correlation with how much the empire had advanced.  The exponential also allowed for a slightly better then flat 0.06 improvement as the distances got larger, though the effect was not huge.  If you experiment with any other calculations, keep in mind that a range of  x<60 is essentially meaningless for this calc, due to the base speed and the floor of 5 turns the calculation only really matters if x is in the hundreds.  

I'll also take a stab at using the speed statistic from the empire's actual ships.

**Comment by Morlic-fo** [comment]
Let me rephrase it then: I do not think the exponent you added is enough to make the new colonization approach on par with using colony ships as it scales too weakly.

I'm not sure why you think x is supposed to be in the hundreds. Maybe you are talking about ShortestDistance which is indeed to be in the hundreds, however I discussed `x=ShortestDistance/80` as stated in the previous comment. x is thus pretty much equivalent to number of turns to build (asymptotically) and thus obviously numbers for x<60 are highly relevant.

I made some mistake though in the calculation and values are off by a factor of 80^0.025 or 11%, but the conclusion remains the same. The effect is an average reduction of 7% of build time which does not compare to a simple ship upgrade which corresponds quickly to 20-50%. 7% is not even a rough approximation of that value but quite subpar. For huge distances, the old, microintensive way thus would quickly outshine it. If we are to keep your current approach, I think it would be a good idea to decrease the exponent some more.

Considering the actual speed seems like a good idea if you manage to script that up. Maybe another idea could be to base the speed on (engine) techs? Let's say, for example, unlocking a new engine tech gives a bonus of 20.

**Comment by Dilvish-fo** [comment]
Ah, my apologies for overlooking/forgetting the normalized distance comment.  My math still works out a bit differently than yours though-- for example, for a total distance of 600, the exponential reduction is over 9%, and at the distances it seems likely Magnate was contemplating swapping to ships for (~15 turns, or probably 1000 -1200+) the exponential is a bit over 10% reduction.  Keep in mind I also gave a starting bonus to speed.  It may still fall a bit short of the actual ships that people would build, and it's certainly an option to consider increasing the exponent.

Another thing to keep in mind is that loading up a colony ship with advanced engines can add quite a bit to its cost, and don't think it was a design requirement that this necessarily match the fastest possible ship at no additional cost.  Basing the reference speed on researched techs is a decent idea, though I don't know that it need to match the absolute strongest engines at no cost.  I guess it would be possible to also have the cost go up in that case.   I think I prefer the 'fastest actually owned ship' approach, but the techs is good to keep in mind as an alternative, and both approaches have their drawbacks.  

Vezzra, this whole thing was really kind of your baby, so if you have any strong preferences I won't argue with you (or, at least not too much :smiley:).

**Comment by Morlic-fo** [comment]
Do you ceil the results in your math? This obviously makes the effect discontinuous in absolut terms and adds fluctuations in the relative effect. That is why I was approximating the result by some step function and so got the average improvement in the first place.

If you have a similar approach, I guess I will have to check my math again, maybe some other step function describes it more accurately.

**Comment by Dilvish-fo** [comment]
I thought we had already established this calculation only mattered for distances above 400 (5 x 80), and even at that low end the reduction is greater than 8.5% (and increases with distance), so to get an average of 7% it seems that at the very least you must have chosen an inappropriate range.

**Comment by Dilvish-fo** [comment]
Also (or instead), I think your normalization doesn't work well here: (x^y)/80 != (x/80)^y

**Comment by Morlic-fo** [comment]
Thats where the mentioned wrong factor came into play. I'll look again into my script to see what else could've went wrong...

**Comment by Dilvish-fo** [comment]
Ah, yes, ok.  What about the ranges?  I hadn't been doing this in a spreadsheet, but I have now and I'm seeing the same results as before.  The spreadsheet does make it easier to play with forms of the adjustment, in the range of 400:2000 an equation of the form x +ax - bx^2 seems to work out pretty decent with a being 1E-3 and b being 3E-4, works up to a 40% reduction.  But anyway I should just try scripting up the actual ship speed statistic, I suppose.

**Comment by Morlic-fo** [comment]
Seems like a random 2 showed up in my formula. Probably mistyped and as the CAS used automatically multiplies numerals and symbols, no error was given. Anyway, multiply my numbers by 2 and we got an agreement. So, sorry for the buzz.

(~12% still may be too low considering naked organics reaching that improvement very early on in the game though, but shouldn't matter much if you're scripting the ship idea anyway now).

**Comment by Vezzra** [comment]
> Vezzra, this whole thing was really kind of your baby, so if you have any strong preferences I won't argue with you (or, at least not too much :smiley:).

I think I'm finally getting around to this. I'm going to take a look at your adjustments, then I'll give you my feedback.

After I've recovered from a bit of math overload I suffered by reading the above exchange, that is :wink:

**Comment by Dilvish-fo** [comment]
ok, just a little update-- the script adjustment using the max speed of any ships owned by the given empire appears to be working (i.e. doesn't crash, gives reasonable results, and doesn't seem to slow things down), although I still need to verify that after building the first super-fast ship the required build times for the buildings do indeed go down.

**Comment by Dilvish-fo** [comment]
Sorry to rush things just as you got involved, Vezzra, but the ship speed statistic seems to work out fine, and I am wanting to get this off my plate, so this last version is committed: 77764c1 
You are of course still invited to make any further changes that look good to you.

**Comment by Vezzra** [comment]
> Sorry to rush things just as you got involved, Vezzra

No problem :wink: However, seems in your hurry you overlooked a serious flaw in your approach. See [this forum thread](http://www.freeorion.org/forum/viewtopic.php?f=15&t=9214&p=76775#p76775), lets continue the discussion there (as you can see, I've also moved your forum post on the topic there).
